subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -655,6 +658,175 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec {
         fails "myTask"
     }
 
+    def "implementation of nested property in Groovy build script is tracked"() {
+        setupTaskClassWithNestedAction()
+        buildFile << """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${originalImplementation}
+            }
+        """
+
+        buildFile.makeOlder()
+
+        when:
+        run 'myTask'
+        then:
+        executedAndNotSkipped(':myTask')
+
+        when:
+        buildFile.text = """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${changedImplementation}
+            }
+        """
```

Comment:
It would be better to test with a static build file, because changing it means the build script classpath changes all the time. We want to test for explicit changes to the action class here instead. Let's declare both actions in the build script, and use a project property (`-Pchanged`) to switch between which one is used.

Reply:
Actually, the classloader for the task does not change, since the task is defined in buildSrc. There are two kinds of changes we track for the implementation classes: the class name and the classloader. If I use something like `-Pchanged`, then the class name changes (`_closure_1`...), if I change the script then the classloader changes.
I went for the classloader changes. Why would you prefer the class name changes?